src/compiler/checker.ts

```diff
@@ -22630,7 +22633,153 @@ namespace ts {
+                     *           if (a[2] !== undefined) {
+                     *               a[1];
+                     *           } else {
+                     *               a[2];  // declared type is union, but here should not be narrowed to never.
+                     *           }
+                     *       }
+                     *   }
+                     */
+                    if (!(declaredType.flags & TypeFlags.Union)) {
+                        return false;
+                    }
+                    //, the types expect primitive should have more than one.
+                    if ((<UnionType>declaredType).types.filter(t => !(t.flags & TypeFlags.Primitive)).length < 2) {
+                        return false;
+                    }
+                    // However, I still use declared type, and put it here. The main aim is for future improving.
+                    // We could deal with 'this' type and use ugly code to deal with Condition 2, then we could avoid use declared type at least for now
+                    // passing all tests and issue #39110/#39114, but does it deserve? Is the code clear enough?
+
+                    // if(isThisTypeParameter(computedType)){
+                    //     return false;
+                    // }
+                    // if(isElementAccessExpression(expr)){
+                    //     // the only purpose of these code is to omit above condition 2.
+                    //     const type = getReducedApparentType(computedType);
+                    //     if (type.flags & TypeFlags.Object) {
+                    //         const resolved = resolveStructuredTypeMembers(<ObjectType>type);
+                    //         if(resolved.stringIndexInfo || resolved.numberIndexInfo){
+                    //             return false;
+                    //         }
+                    //     }
+                    // }
+                    return true;
+                }
+
+                // here are all possiable situations:
+                // 1. nullable | nonNullable1
+                // 2. nullable | nonNullable1 | nonNullable2...
+                // 3. nonNullable1 | nonNullable2...
+                // if root contains nullable type, for situation 1, we do not need to narrow.
+                const computedTypeIsNullable = maybeTypeOfKind(computedType, TypeFlags.Nullable);
+                return isTypeArrayDiscriminant(reachableFinalType.concat(rootNullable), computedTypeIsNullable);
+
+                // main part is copied from function createUnionOrIntersectionProperty
+                function isTypeArrayDiscriminant(types: Type[], nullableFlag: boolean) {
+                    let checkFlags = 0;
+                    let firstType: Type | undefined;
+                    for (const type of types) {
+                        if (nullableFlag && (type.flags & TypeFlags.Nullable)) {
+                            continue;
+                        }

```

Comment:
I believe the new problem is that this prevents you from being able to discriminate between `undefined` and another literal, because we never set `HasNonUniformType`.